![image info](https://user-images.githubusercontent.com/91945811/146929832-e78ff1f4-2739-41b1-acd3-019a39b6a42c.png)

# Kaffeebohnen Projekt

Willkommen zu unserem Data-Science-Projekt! Heute nehmen wir dich mit auf eine unterhaltsame Reise durch die chaotische, aber liebenswerte Welt der **Kaffeebohnen GmbH**. Dieses familiengeführte Unternehmen ist bekannt für seinen hervorragenden Kaffee – aber weniger für seine exzellente Datenverwaltung.

### Die manuelle Datenaufnahme bei der Kaffeebohnen GmbH

Bei der Kaffeebohnen GmbH ist die Datenpflege Familiensache. **Oma Erna**, das Herz des Unternehmens, kümmert sich persönlich um die Aufzeichnung aller Verkaufsdaten. Bewaffnet mit ihrem treuen alten Laptop (Baujahr 1999) und einer Tasse ihres Lieblingskaffees tippt sie fleißig jede Bestellung in Excel ein.

Doch Oma Erna ist eine Frau der Kreativität:

- **Spaltennamen nach Lust und Laune**: Was letztes Jahr noch "Produkt" hieß, nennt sich dieses Jahr "Kaffeesorte" und nächstes Jahr vielleicht "Artikelbezeichnung".
- **Individuelle Formate**: Datumsangaben variieren zwischen "31.12.2020", "2020/12/31" und "31-12-20". Manchmal schreibt sie das Datum sogar als "Silvester 2020".
- **Tippfehler und künstlerische Freiheit**: Aus "Espresso" wird schnell mal "Expresso" oder "Espreso". Und wer könnte ihr verdenken, dass sie "Cappuccino" als "Kapuccino" schreibt?

Natürlich meint es Oma Erna nur gut, aber ihre einzigartige Art der Dateneingabe stellt uns vor einige Herausforderungen.


### Unsere Mission

In diesem Projekt werden wir:

- Die Daten der letzten Jahre aus verschiedenen Excel-Dateien laden.
- Die unterschiedlichen Spaltennamen vereinheitlichen.
- Tippfehler bereinigen und Namen standardisieren.
- Datenformate korrigieren und konsistent machen.
- Fehlende Werte und Duplikate behandeln.


Mach dich bereit für eine spannende Datenrettungsaktion! Mit Python als unserem Werkzeug und einem Sinn für Humor werden wir Ordnung in das Datenchaos der Kaffeebohnen GmbH bringen.

---

**Disclaimer:** Dieses Projekt ist rein fiktiv, und die verwendeten Daten sind vollständig erfunden. Dennoch sind die dargestellten Probleme und Herausforderungen typisch und treten in der Praxis häufig auf... und vergiss nicht, zwischendurch eine Tasse Kaffee zu genießen – das hat bei Oma Erna schließlich auch immer geholfen!

---

In [3]:
import pandas as pd

## Daten laden

#### 1. Lese *Kaffeebohnen_2019.xlsx* korrekt ein

Öffne die Excel-Dateien zuerst aus dem Ordner *KaffeeDaten* manuell in Excel, um Folgendes zu überprüfen:

- Datenstruktur: Gibt es einheitliche Muster in den Tabellen?
- Header: Gibt es mehrere Headerzeilen oder eine korrekte Kopfzeile?
- Bereich: Liegen die relevanten Daten in einem spezifischen Bereich?

Informiere dich über die Parameter der Funktion `read_excel()` und wähle gezielt passende Einstellungen aus, um den relevanten Datenbereich effizient zu laden.

- Lade mit `pandas` alle drei Dateien in den DataFrames `df19`, `df20` und `df21`.

- Stelle sicher, dass bei allen DataFrames die Header-Zeile stimmt (z.B. mit `skiprows` oder `header`) und die gleiche Anzahl von Spalten geladen wird (z.B. mit `usecols`).

In [4]:
df19 = pd.read_excel("KaffeeDaten/Kaffeebohnen_2019.xlsx")

df20 = pd.read_excel("KaffeeDaten/Kaffeebohnen_2020.xlsx")

df21 = pd.read_excel("KaffeeDaten/Kaffeebohnen_2021.xlsx")

In [5]:
display(df19)
display(df20)
display(df21)

,Verkaufszahlen im Jahr 2019,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5
0,NaN,NaN,NaN,NaN,NaN,NaN
1,Kaffee Sorte,€,Stückzahl,Kauf Datum,Kunde bestellt aus:,Kunden-ID
2,Latte Macchiato,3.5,8,2019-02-08 00:00:00,Austria,NaN
3,Expresso,2,3,2019-10-14 00:00:00,Deutschland,1022
4,Amerikana,2.5,1,2019-04-26 00:00:00,Austria,1011
...,...,...,...,...,...,...
58,Kapuccino,3,8,2019-07-09 00:00:00,D,0
59,Latte Macchiato,3.5,1,2019-12-11 00:00:00,GER,1076
60,Latte Macchiato,3.5,2,2019-09-24 00:00:00,Deutschland,1025
61,Expresso,2,1,2019-06-17 00:00:00,Germany,1045


,Excel Liste für das Jahr 2020,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Kaffee,Preis,Stück,Datum,Land,Kunden-ID,NaN
2,Amerikana,2.5,2,2020-01-18 00:00:00,Deutschland,1002,NaN
3,Americano,2.5,3,2020-06-29 00:00:00,D,1081,NaN
4,Kapuccino,3,3,2020-12-01 00:00:00,Austria,1038,NaN
...,...,...,...,...,...,...,...
83,Expresso,2,1,Neujahr,Österreich,1065,NaN
84,Espreso,2,Storniert,2020-11-07 00:00:00,D,1093,Kunde hat storniert
85,Cappuccino,3,1,2020-03-15 00:00:00,D,1064,NaN
86,Cappuccino,3,1,2020-07-14 00:00:00,GER,1056,NaN


,"Lieber Olli, anbei die Übersicht für das Jahr 2021 :)",Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Sorte,Preis in €,Mengenangabe,Kaufdatum,Land,Kundennr,NaN
2,Espresso,2,4,2021-10-23 00:00:00,GER,1029,NaN
3,Kapuccino,3,3,2021-10-18 00:00:00,GER,1033,NaN
4,Expresso,"2,0",4,2021-02-14 00:00:00,GER,"1068, Kunde hat Relkamiert!",NaN
...,...,...,...,...,...,...,...
90,Cappuccino,3,1,2021-06-17 00:00:00,Österreich,1054,NaN
91,Amerikana,"2,5",1,2021-12-29 00:00:00,D,1087,NaN
92,Espreso,2,4,2021-01-06 00:00:00,Deutschland,1022,NaN
93,Espreso,2,3,2021-05-10 00:00:00,Germany,1073,NaN


In [6]:
print(f"{df19['Unnamed: 2'].unique()}")
print(f"{df20['Unnamed: 2'].unique()}")
print(f"{df21['Unnamed: 2'].unique()}")

[nan 'Stückzahl' 8 3 1 4 100 2 0]
[nan 'Stück' 2 3 1 100 4 300 8 'Storniert']
[nan 'Mengenangabe' 4 3 1 2 8 '?']


In [7]:
print(f"{df20['Unnamed: 6'].unique()}")
print(f"{df21['Unnamed: 6'].unique()}")

[nan 'Kunde hat storniert']
[nan 'Kundennummer verloren!']


#### 2. Tabellen zusammenführen

Alle Excel-Dateien sind leider nicht einheitlich aufgebaut: Die Tabelle enthalten genau 6 Spalten, die aber unterschiedlich benannt sind, aber den gleichen Inhalt haben. 

- Verwende einheitliche Spaltennamen für alle drei DataFrames.

- Führe alle Excel-Dateien in einem einzigen DataFrame mit `pd.concat` zusammen. 

- Validiere, dass die Anzahl von Zeilen nach dem `pd.concat` auch passt.

Der Vorteil der Zusammenführung ist, dass weitere Bereinigungsmaßnahmen anschließend nur einmalig durchgeführt werden müssten.

In [8]:
df_list = [df19, df20, df21]
df19.name = "df19"
df20.name = "df20"
df21.name = "df21"

df_copy_list = []

for df in df_list:
    print(f"{df.name} has a shape of {df.shape}.")
    df_copy = df.iloc[
        2:, :6
    ]  # remove the frist two rows of each dataframe, only takes the first 6 columns
    df_copy.name = df.name + "_copy"
    print(f"{df_copy.name} has a shape of {df_copy.shape}.")
    df_copy.columns = ["produkt", "preis", "menge", "datum", "kunde_bestellt_land", "kunden_id"]
    df_copy_list.append(df_copy)

df19 has a shape of (63, 6).
df19_copy has a shape of (61, 6).
df20 has a shape of (88, 7).
df20_copy has a shape of (86, 6).
df21 has a shape of (95, 7).
df21_copy has a shape of (93, 6).


In [9]:
df_all_years = pd.concat(df_copy_list, axis=0, ignore_index=True)

In [10]:
display(df_all_years)

,produkt,preis,menge,datum,kunde_bestellt_land,kunden_id
0,Latte Macchiato,3.5,8,2019-02-08 00:00:00,Austria,NaN
1,Expresso,2,3,2019-10-14 00:00:00,Deutschland,1022
2,Amerikana,2.5,1,2019-04-26 00:00:00,Austria,1011
3,Kapuccino,3,3,2019-02-28 00:00:00,Österreich,1012
4,Amerikana,2.5,4,2019-10-06 00:00:00,DE,1082
...,...,...,...,...,...,...
235,Cappuccino,3,1,2021-06-17 00:00:00,Österreich,1054
236,Amerikana,"2,5",1,2021-12-29 00:00:00,D,1087
237,Espreso,2,4,2021-01-06 00:00:00,Deutschland,1022
238,Espreso,2,3,2021-05-10 00:00:00,Germany,1073


In [11]:
df_all_years.shape

(240, 6)

#### 3. Verschaffe dir eine Übersicht über die Daten

- Gebe eine kompakte Zusammenfassung der Struktur des DataFrames, einschließlich Spaltennamen, Datentypen und Anzahl der nicht-leeren Werte

- Liefere grundlegende statistische Kennzahlen und erste Einblicke in die Verteilung der Daten zu erhalten

In [12]:
df_all_years.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240 entries, 0 to 239
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   produkt              222 non-null    object
 1   preis                222 non-null    object
 2   menge                222 non-null    object
 3   datum                222 non-null    object
 4   kunde_bestellt_land  222 non-null    object
 5   kunden_id            211 non-null    object
dtypes: object(6)
memory usage: 11.4+ KB


In [13]:
df_all_years.describe()

,produkt,preis,menge,datum,kunde_bestellt_land,kunden_id
count,222,222,222,222,222,211
unique,9,8,10,186,8,87
top,Kapuccino,2,1,2020/02/30,D,1093
freq,31,67,72,5,41,7


## Datenbereinigung

#### 4. Entfernen von leeren Zeilen

Wir entfernen zuerst ausschließlich die Zeilen, die in **allen** Spalten fehlende Werte enthalten, um sicherzustellen, dass teilweise ausgefüllte Datensätze erhalten bleiben.

Überprüfe nun erneut die `pd.info()`, um sicher zu sein, dass jetzt keine leeren Zeilen mehr im DataFrame enthalten sind.

In [14]:
df_all_years = df_all_years.dropna(how="all")

In [15]:
df_all_years.info()

<class 'pandas.core.frame.DataFrame'>
Index: 222 entries, 0 to 239
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   produkt              222 non-null    object
 1   preis                222 non-null    object
 2   menge                222 non-null    object
 3   datum                222 non-null    object
 4   kunde_bestellt_land  222 non-null    object
 5   kunden_id            211 non-null    object
dtypes: object(6)
memory usage: 12.1+ KB


#### 5. Anzeigen von Zeilen ohne Wert

Der nächste Schritt ist, die Zeilen anzuzeigen, in denen eine Zelle keinen Wert enthält. Dies hilft uns, potenzielle fehlende Informationen zu identifizieren, bevor wir entscheiden, wie wir damit umgehen. Mache dir Gedanken darüber, ob solche Zeilen entfernt werden sollten. 

In diesem Fall sehen wir, dass von den 222 Zeilen in der Spalte `Kunden_ID` nur 211 davon *non-null* sind. Zeige diese 11 Zeilen auf der Konsole.

In [16]:
df_all_years.isnull().sum()

produkt                 0
preis                   0
menge                   0
datum                   0
kunde_bestellt_land     0
kunden_id              11
dtype: int64

In [17]:
df_all_years[df_all_years["kunden_id"].isna()]

,produkt,preis,menge,datum,kunde_bestellt_land,kunden_id
0,Latte Macchiato,3.5,8,2019-02-08 00:00:00,Austria,NaN
23,Kapuccino,3,3,2019-12-05 00:00:00,Österreich,NaN
26,Cappuccino,3,2,2019-01-15 00:00:00,Österreich,NaN
29,Espresso,2,2,2019-01-08 00:00:00,D,NaN
85,Expresso,2,2,2020-01-13 00:00:00,Germany,NaN
103,Latte Macchiato,3.5,2,2020-05-19 00:00:00,GER,NaN
111,Kapuccino,3,2,2020-02-11 00:00:00,GER,NaN
117,Kapuccino,3,1,2020-07-16 00:00:00,AUT,NaN
151,Kapuccino,3,3,2021-02-25 00:00:00,GER,NaN
227,Latte Macchiato,3.5,1,2021-08-14 00:00:00,Austria,NaN


#### 6. Überprüfung der Spalten auf Tippfehler und Korrektur

Jetzt geht es ans Eingemachte. Um mögliche Tippfehler in den Daten aufzuspüren und zu beheben, solltest du die folgenden Schritte durchführen:

**Identifikation einzigartiger Werte:** 

Zeige zunächst alle eindeutigen Werte der Spalte `Produkt` an. Dadurch erhältst du einen Überblick über die vorhandenen Einträge und erkennst mögliche Tippfehler oder Unstimmigkeiten. Rechne gerne auch die Häufigkeit dieser Werte mit `value_counts` aus.

Prüfe die einzigartigen Einträge auf Unregelmäßigkeiten, wie z. B. falsche Schreibweisen, unvollständige Begriffe oder ungewöhnliche Sonderzeichen. 

Entwickele eine Strategie, um diese Fehler zu korrigieren, z.B. durch Ersetzen häufiger Tippfehler, Harmonisierung von Synonymen oder Anpassung an einen definierten Standard.

In [18]:
df_all_years.info()

<class 'pandas.core.frame.DataFrame'>
Index: 222 entries, 0 to 239
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   produkt              222 non-null    object
 1   preis                222 non-null    object
 2   menge                222 non-null    object
 3   datum                222 non-null    object
 4   kunde_bestellt_land  222 non-null    object
 5   kunden_id            211 non-null    object
dtypes: object(6)
memory usage: 12.1+ KB


In [19]:
df_all_years["produkt"].unique()

array(['Latte Macchiato', 'Expresso', 'Amerikana', 'Kapuccino',
       'Espresso', 'Latte Machiato', 'Americano', 'Espreso', 'Cappuccino'],
      dtype=object)

In [20]:
correct_kaffee = {
    "Expresso": "Espresso",
    "Amerikana": "Americano",
    "Kapuccino": "Cappuccino",
    "Latte Machiato": "Latte Macchiato",
    "Espreso": "Espresso",
}

In [33]:
df_all_years["produkt"] = df_all_years["produkt"].replace(correct_kaffee)
df_all_years["produkt"] = df_all_years["produkt"].astype("string")

In [22]:
df_all_years["produkt"].unique()

array(['Latte Macchiato', 'Espresso', 'Americano', 'Cappuccino'],
      dtype=object)

**Spalten Typ anpassen**

- Zeige erstmal die `value_counts` für die Spalten `Preis`, `Menge` und `Datum`.

- In der Spalte `Preis` musst du noch das `,` mit einem `.` ersetzen. Dafür kannst du die Syntax `df['Preis'].astype(str).str.replace(",", ".")` nutzen.

- Konvertiere die Spalten ins passende Datenformat mit `to_numeric` und `to_datetime`.

- Nutze den Parameter `errors='coerse'`, um fehlerhafte Werte mit `NaN` zu erstetzen.

- Zeige am Ende die Zeilen für das Datum, wo man das nicht automatisch in einem einheitlichen Format bringen kann.

In [63]:
print("Value counts for preis:")
print(df_all_years["preis"].value_counts())

print("\nValue counts for menge:")
print(df_all_years["menge"].value_counts())

print("\nValue counts for datum:")
print(df_all_years["datum"].value_counts())

Value counts for preis:
preis
2.0    61
3.0    51
3.5    43
2.5    42
Name: count, dtype: int64

Value counts for menge:
menge
1      62
2      53
3      46
4      26
8       7
100     2
300     1
Name: count, dtype: int64

Value counts for datum:


KeyError: 'datum'

In [26]:
df_all_years["preis"] = df_all_years["preis"].astype(str).str.replace(",", ".")
df_all_years["preis"] = pd.to_numeric(df_all_years["preis"], errors="coerce")

In [38]:
index_to_drop = df_all_years[df_all_years["menge"].isin(["Storniert", "?", 0])].index
df_all_years.drop(index_to_drop, inplace=True)
df_all_years["menge"] = pd.to_numeric(df_all_years["menge"], errors="coerce")

In [40]:
print("\nValue counts for menge:")
print(df_all_years["menge"].value_counts())


Value counts for menge:
menge
1      72
2      56
3      50
4      30
8       8
100     2
300     1
Name: count, dtype: int64


In [41]:
df_all_years.info()

<class 'pandas.core.frame.DataFrame'>
Index: 219 entries, 0 to 239
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   produkt              219 non-null    string 
 1   preis                219 non-null    float64
 2   menge                219 non-null    int64  
 3   datum                219 non-null    object 
 4   kunde_bestellt_land  219 non-null    object 
 5   kunden_id            208 non-null    object 
dtypes: float64(1), int64(1), object(3), string(1)
memory usage: 12.0+ KB


  
**Duplikate**

Zeige erstmal alle Duplikate auf der Konsole, bevor du diese mit `drop_duplicates` entfernst.

In [42]:
df_all_years[df_all_years.duplicated()].sort_values(by=["produkt", "datum", "kunden_id"])

,produkt,preis,menge,datum,kunde_bestellt_land,kunden_id
236,Americano,2.5,1,2021-12-29 00:00:00,D,1087
144,Cappuccino,3.0,1,2020-03-15 00:00:00,D,1064
145,Cappuccino,3.0,1,2020-07-14 00:00:00,GER,1056
60,Espresso,2.0,2,2019-04-14 00:00:00,D,1057
59,Espresso,2.0,1,2019-06-17 00:00:00,Germany,1045
146,Espresso,2.0,4,2020-09-01 00:00:00,AUT,1025
238,Espresso,2.0,3,2021-05-10 00:00:00,Germany,1073
58,Latte Macchiato,3.5,2,2019-09-24 00:00:00,Deutschland,1025


In [43]:
df_all_years.shape

(219, 6)

In [44]:
df_all_years.drop_duplicates(inplace=True)

In [45]:
df_all_years.shape

(211, 6)

**Übersicht über die Daten**:

Du hast die Datenquelle erfolgreich bereinigt. Gebe nun wieder eine kompakte Zusammenfassung der Struktur des DataFrames, einschließlich Spaltennamen, Datentypen und Anzahl der nicht-leeren Werte und liefere grundlegende statistische Kennzahlen.

In [46]:
df_all_years["datum_cleaned"] = pd.to_datetime(df_all_years["datum"], errors="coerce")

In [47]:
df_all_years[df_all_years["datum_cleaned"].isna()]

,produkt,preis,menge,datum,kunde_bestellt_land,kunden_id,datum_cleaned
16,Espresso,2.0,4,15-15-2019,Deutschland,1032,NaT
31,Cappuccino,3.0,1,15-15-2019,Austria,1047,NaT
38,Latte Macchiato,3.5,2,2020/02/30,Germany,1057,NaT
93,Latte Macchiato,3.5,1,15-15-2019,DE,1088,NaT
106,Espresso,2.0,4,Osterzeit,GER,1024,NaT
114,Espresso,2.0,1,2020/02/30,AUT,1064,NaT
118,Cappuccino,3.0,3,2020/02/30,Deutschland,1091,NaT
119,Americano,2.5,4,2020/02/30,D,1040,NaT
142,Espresso,2.0,1,Neujahr,Österreich,1065,NaT
162,Espresso,2.0,1,Datum nicht lesbar,Austria,1058,NaT


In [48]:
df_all_years = df_all_years.dropna(subset=["datum_cleaned"])

In [52]:
df_all_years = df_all_years.drop("datum", axis=1)

In [53]:
df_all_years.shape

(197, 6)

In [54]:
df_all_years.dtypes

produkt                string[python]
preis                         float64
menge                           int64
kunde_bestellt_land            object
kunden_id                      object
datum_cleaned          datetime64[ns]
dtype: object

In [55]:
df_all_years["kunde_bestellt_land"].unique()

array(['Austria', 'Deutschland', 'Österreich', 'DE', 'AUT', 'D', 'GER',
       'Germany'], dtype=object)

In [59]:
correct_land = {
    "Austria": "Österreich",
    "DE": "Deutschland",
    "AUT": "Österreich",
    "D": "Deutschland",
    "GER": "Deutschland",
    "Germany": "Deutschland",
}

df_all_years["kunde_bestellt_land"] = df_all_years["kunde_bestellt_land"].replace(correct_land)
df_all_years["kunde_bestellt_land"] = df_all_years["kunde_bestellt_land"].astype("string")

In [60]:
df_all_years["kunde_bestellt_land"].unique()

<StringArray>
['Österreich', 'Deutschland']
Length: 2, dtype: string

In [64]:
df_all_years["kunden_id"].value_counts()

kunden_id
0       6
1087    6
1093    6
1025    5
1012    5
       ..
1013    1
1059    1
1014    1
1018    1
1072    1
Name: count, Length: 85, dtype: int64

In [65]:
df_all_years["kunden_id"].unique()

array([nan, 1022, 1011, 1012, 1082, 1044, 1050, 1005, 1055, 1025, 1098,
       1000, 1027, 1024, 1062, 1063, 1023, 1071, 1070, 1081, 1094, 1018,
       1003, 1014, 1059, 1069, 1004, 1013, 1077, 1065, 1057, 1073, 0,
       1043, 1015, 1051, 1041, 1089, 1045, 1035, 1048, 1036, 1042, 1076,
       1002, 1038, 1093, 1010, 1016, 1061, 1074, 1087, 1091, 1080, 1046,
       1056, 1037, 1058, 1030, 1040, 1052, 1095, 1064, 1001, 1049, 1060,
       1084, 1020, 1031, 1079, 1075, 1029, 1033,
       '1068, Kunde hat Relkamiert!', 1034, 1090, 1099, 1039, 1006, 1088,
       1019, 1054, 1086, 1017, 1085, 1072], dtype=object)

In [66]:
df_all_years.loc[df_all_years["kunden_id"] == "1068, Kunde hat Relkamiert!", "kunden_id"] = 1068

In [68]:
df_all_years["kunden_id"] = df_all_years["kunden_id"].astype("Int64")

In [69]:
df_all_years.dtypes

produkt                string[python]
preis                         float64
menge                           int64
kunde_bestellt_land    string[python]
kunden_id                       Int64
datum_cleaned          datetime64[ns]
dtype: object

**Identifikation von Ausreißern:** 

Überlege, ob einige ungewöhnliche Einträge tatsächlich valide sind oder ob es sich um Fehler handelt. Vergleiche z.B. mit der `describe` Tabelle, welche Werte deutlich über den 75%-Wert liegen.

In [72]:
df_all_years.describe()

,preis,menge,kunden_id,datum_cleaned
count,197.000000,197.000000,186.0,197
mean,2.692893,4.903553,1015.865591,2020-08-02 14:59:05.177665024
min,2.000000,1.000000,0.0,2019-01-08 00:00:00
25%,2.000000,1.000000,1023.0,2019-12-12 00:00:00
50%,2.500000,2.000000,1046.0,2020-08-24 00:00:00
75%,3.000000,3.000000,1074.0,2021-04-15 00:00:00
max,3.500000,300.000000,1099.0,2021-12-30 00:00:00
std,0.569954,23.344707,188.158985,NaN


**Bereinigte Tabelle extrahieren**:

Speichere die bereinigte Tabelle als Excel-Datei mit `to_excel` wieder ab.

In [ ]:
df_all_years.to_excel("Kaffeebohnen_all.xlsx")